### PR TITLE
Extend LSTM, GRU and SimpleRNN from zoo.keras.Recurrent and fix the serde issue

### PR DESCRIPTION
**Pull request by zhichao-li**
https://github.com/intel-analytics/analytics-zoo/issues/723

### CODE REVIEW

**Comment by dding3** [comment]
Do we have UT for serialization test?

**Comment by dding3** [comment]
If we remove InternalRecurrent, one thing is the new apis added in InternalRecurrent cannot be moved to zoo.keras.recurrent as the apis needed to access some variables that are protected.

Can we keep InternalRecurrent and copy nn.Recurrent loadModule/serializeModule into InternalRecurrent to fix the serialize problem? Looks like the variables needed to be serialized are protected and InternalRecurrent should be able to access them?

**Comment by zhichao-li** [comment]
One of my concern is it would introduce tons of duplicated code. You can also try to use reflection for those protected methods if there's no concern on the performance.

**Comment by zhichao-li** [comment]
One of my concern is it would introduce tons of duplicated code. You can also try to use reflection for those protected methods if there's no concern on the performance.

**Comment by dding3** [comment]
Thanks for your update! The performance is not the concern here as load/save only happened handful times and should not affect training/inference.

BTW, do we have unit tests to test serialize for zoo.keras rnn, like LSTM?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/LSTM.scala [hunk 1]
```diff
@@ -44,29 +44,29 @@ import scala.reflect.ClassTag
...
  * @param uRegularizer An instance of [[Regularizer]], applied the recurrent weights matrices.
  *                     Default is null.
  * @param bRegularizer An instance of [[Regularizer]], applied to the bias. Default is null.
- * @param inputShape A Single Shape, does not include the batch dimension.
+ * @param mInputShape A Single Shape, does not include the batch dimension.
  * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now.
  */
 class LSTM[T: ClassTag](
-   override val outputDim: Int,
+   var outputDimension: Int,
```
[the remark this answers is not in the document]
Comment (zhichao-li) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/LSTM.scala]: It's a work around for a bug on the SerDe in BigDL.

**Comment by zhichao-li** [comment]
There are already serde test to cover those rnn layers.

**Comment by dding3** [comment]
LGTM.

**Comment by zhichao-li** [comment]
Jenkins passed. ZOO-PR-Validation/1253/

**Comment by hkvision** [comment]
1. Have you removed the excluded ser test of InternalRecurrent from https://github.com/intel-analytics/analytics-zoo/blob/b850fac55fde749ac0c917fc2e2602b9513c82ed/zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/serializer/SerializerSpecHelper.scala#L38 ?
2. Why aren't you doing the same changes for SimpleRNN (e.g. change val to var for some arguments)

**Pull request merged**